In [1]:
import pandas as pd
import logging

# Log fayl manzili
log_path = r"C:\Users\Rasulbek907\Desktop\Project_MP\Log\tuning.log"

# Log sozlamalari
logging.basicConfig(
    filename=log_path,
    filemode='a',  # Append mode
    format='%(asctime)s - %(levelname)s - %(message)s',
    level=logging.INFO
)

try:
    logging.info("CSV fayl o'qilmoqda:...")
    df = pd.read_csv(r"C:\Users\Rasulbek907\Desktop\Project_MP\Data\Feature_Selection\Filtered_Features.csv")
    logging.info(f"Fayl muvaffaqiyatli o'qildi. Satırlar soni: {len(df)} ustunlar soni: {len(df.columns)}")
except Exception as e:
    logging.error(f"CSV faylni o'qishda xatolik: {e}")

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4993 entries, 0 to 4992
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0.1           4993 non-null   int64  
 1   Unnamed: 0             4993 non-null   int64  
 2   name_length            4993 non-null   float64
 3   is_organic             4993 non-null   float64
 4   quantity_value         4993 non-null   float64
 5   quantity_unit_encoded  4993 non-null   float64
 6   category_depth         4993 non-null   float64
 7   country_count          4993 non-null   float64
 8   product_age_days       4993 non-null   float64
 9   created_month          4993 non-null   float64
 10  main_category_encoded  4993 non-null   float64
 11  nova_group             4993 non-null   float64
dtypes: float64(10), int64(2)
memory usage: 468.2 KB


In [3]:
import sys
import os

# Source papkasini yo'lga qo'shamiz
source_path = os.path.abspath("../Source")
if source_path not in sys.path:
    sys.path.append(source_path)

print("Python executable:", sys.executable)
print("Source path added:", source_path)

# Endi modulni import qilamiz
from preprosessing import Cleaner, Encoder, Scaler

Python executable: c:\Users\Rasulbek907\Desktop\Project_MP\venv\Scripts\python.exe
Source path added: c:\Users\Rasulbek907\Desktop\Project_MP\Source


In [4]:
cleaner = Cleaner(df)
df = cleaner.tozala().get_df()

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4993 entries, 0 to 4992
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0.1           4993 non-null   int64  
 1   Unnamed: 0             4993 non-null   int64  
 2   name_length            4993 non-null   float64
 3   is_organic             4993 non-null   float64
 4   quantity_value         4993 non-null   float64
 5   quantity_unit_encoded  4993 non-null   float64
 6   category_depth         4993 non-null   float64
 7   country_count          4993 non-null   float64
 8   product_age_days       4993 non-null   float64
 9   created_month          4993 non-null   float64
 10  main_category_encoded  4993 non-null   float64
 11  nova_group             4993 non-null   float64
dtypes: float64(10), int64(2)
memory usage: 468.2 KB


In [6]:
encoder = Encoder(df)
df = encoder.encodla().get_df()

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4993 entries, 0 to 4992
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0.1           4993 non-null   int64  
 1   Unnamed: 0             4993 non-null   int64  
 2   name_length            4993 non-null   float64
 3   is_organic             4993 non-null   float64
 4   quantity_value         4993 non-null   float64
 5   quantity_unit_encoded  4993 non-null   float64
 6   category_depth         4993 non-null   float64
 7   country_count          4993 non-null   float64
 8   product_age_days       4993 non-null   float64
 9   created_month          4993 non-null   float64
 10  main_category_encoded  4993 non-null   float64
 11  nova_group             4993 non-null   float64
dtypes: float64(10), int64(2)
memory usage: 468.2 KB


In [8]:
scaler = Scaler(df, target_col="nova_group")
scaled_df = scaler.scaling_qil().get_df()

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4993 entries, 0 to 4992
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0.1           4993 non-null   int64  
 1   Unnamed: 0             4993 non-null   int64  
 2   name_length            4993 non-null   float64
 3   is_organic             4993 non-null   float64
 4   quantity_value         4993 non-null   float64
 5   quantity_unit_encoded  4993 non-null   float64
 6   category_depth         4993 non-null   float64
 7   country_count          4993 non-null   float64
 8   product_age_days       4993 non-null   float64
 9   created_month          4993 non-null   float64
 10  main_category_encoded  4993 non-null   float64
 11  nova_group             4993 non-null   float64
dtypes: float64(10), int64(2)
memory usage: 468.2 KB


# Logistic Regression

In [10]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

# X va y ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# Train / Test ajratish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Model yaratish
log_model = LogisticRegression(max_iter=500, random_state=42)
log_model.fit(X_train, y_train)

# Bashorat
y_pred = log_model.predict(X_test)

# Baholash
print("📊 Logistic Regression natijalari:")
print("Accuracy:", accuracy_score(y_test, y_pred))

📊 Logistic Regression natijalari:
Accuracy: 0.5425425425425425


c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 500 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=500).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# Decision Tree Classifier 

In [11]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# X va y ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# Train / Test ajratish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Model yaratish (default sozlamalar bilan)
dt_model = DecisionTreeClassifier()

# O‘qitish
dt_model.fit(X_train, y_train)

# Bashorat
y_pred = dt_model.predict(X_test)

# Natijalar
print("🌳 Decision Tree Classifier natijalari:")
print("Accuracy:", accuracy_score(y_test, y_pred))

🌳 Decision Tree Classifier natijalari:
Accuracy: 0.46846846846846846


# Random Forest Classifier

In [12]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

X = df.drop(columns="nova_group")
y = df["nova_group"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

rf_model = RandomForestClassifier()  # default sozlamalar
rf_model.fit(X_train, y_train)

y_pred = rf_model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))


Accuracy: 0.5825825825825826


# HYPERPARAMETER TUNING 

# MANUAL SEARCH 

# Description: Adjust hyperparameters by intuition or trial-and-error. # Pros: Simple # Cons: Inefficient, may miss optimal values

#  RandomForestClassifier + Manual Search

In [13]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


# 🧩 Model (qo‘lda hyperparameterlar bilan)
rf_manual = RandomForestClassifier(
    n_estimators=2,    # daraxtlar soni
    max_depth=5,       # maksimal chuqurlik
    random_state=42
)

# O‘qitish
rf_manual.fit(X_train, y_train)

# Bashorat
y_pred_manual = rf_manual.predict(X_test)

# Baholash
r2_manual = accuracy_score(y_test, y_pred_manual)
print("🌲 Random Forest (Manual Search) Accuracy:", r2_manual)


🌲 Random Forest (Manual Search) Accuracy: 0.5455455455455456


#  DecisionTreeClassifier + Manual Search

In [14]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score


# 🌳 Model — qo‘lda tanlangan hyperparametrlar bilan
dt_manual = DecisionTreeClassifier(
    criterion="entropy",   # yoki "gini"
    max_depth=5,           # maksimal chuqurlik
    min_samples_split=4,   # bo‘linish uchun minimal namunalar
    random_state=42
)

# O‘qitish
dt_manual.fit(X_train, y_train)

# Bashorat
y_pred_manual = dt_manual.predict(X_test)

# Baholash
acc_manual = accuracy_score(y_test, y_pred_manual)
print("🌳 Decision Tree (Manual Search) Accuracy:", acc_manual)


🌳 Decision Tree (Manual Search) Accuracy: 0.5455455455455456


#  LogisticRegression + Manual Search


In [15]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score



log_manual = LogisticRegression(
    solver="liblinear",   # kichik datasetlar uchun yaxshi
    penalty="l2",         # regularizatsiya turi ("l1" yoki "l2")
    C=0.7,                # regularizatsiya kuchi (kichik bo‘lsa — ko‘proq jazo)
    max_iter=500,         # iteratsiyalar soni
    random_state=42
)

# Modelni o‘qitish
log_manual.fit(X_train, y_train)

# Bashorat
y_pred_manual = log_manual.predict(X_test)

# Baholash
acc_manual = accuracy_score(y_test, y_pred_manual)
print("📊 Logistic Regression (Manual Search) Accuracy:", acc_manual)


📊 Logistic Regression (Manual Search) Accuracy: 0.5325325325325325


c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:1296: FutureWarning: Using the 'liblinear' solver for multiclass classification is deprecated. An error will be raised in 1.8. Either use another solver which supports the multinomial loss or wrap the estimator in a OneVsRestClassifier to keep applying a one-versus-rest scheme.
  warnings.warn(


In [21]:
import pandas as pd

# 📊 Modellar natijalari
results = pd.DataFrame({
    'Model nomi': [
        '📊 Logistic Regression',
        '🌳 Decision Tree Classifier',
        '🌲 Random Forest Classifier',
        '📊 Logistic Regression (Manual Search)',
        '🌳 Decision Tree (Manual Search)',
        '🌲 Random Forest (Manual Search)'
    ],
    'Aniqlik (Accuracy)': [
        0.5425425425425425,
        0.46846846846846846,
        0.5825825825825826,
        0.5325325325325325,
        0.5455455455455456,
        0.5455455455455456
    ]
})

# 🔄 O'sish yoki kamayishni aniqlash uchun avvalgi qator bilan farqni topamiz
results['O‘zgarish'] = results['Aniqlik (Accuracy)'].diff().fillna(0)

# 🎨 Rang beruvchi funksiya
def color_change(val):
    if val > 0:
        color = 'green'
    elif val < 0:
        color = 'red'
    else:
        color = 'black'
    return f'color: {color}; font-weight: bold'

# 🧾 Natijani chiroyli chiqarish
styled = results.style.format({'Aniqlik (Accuracy)': '{:.3f}'}) \
                      .applymap(color_change, subset=['O‘zgarish'])

styled


C:\Users\Rasulbek907\AppData\Local\Temp\ipykernel_13996\2719313647.py:38: FutureWarning: Styler.applymap has been deprecated. Use Styler.map instead.
  .applymap(color_change, subset=['O‘zgarish'])


# GRID SEARCH 

# RandomForestClassifier + Grid Search

In [16]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report

rf = RandomForestClassifier(random_state=42)

# 🔍 Grid Search uchun parametrlar to‘plami
param_grid = {
    'n_estimators': [50, 100, 150],     # daraxtlar soni
    'max_depth': [None, 5, 10, 15],     # daraxt chuqurligi
    'min_samples_split': [2, 5, 10],    # bo‘linish uchun minimal namunalar
    'min_samples_leaf': [1, 2, 4],      # barg tugunlari uchun minimal namunalar
    'criterion': ['gini', 'entropy']    # ajratish mezoni
}

# 🧠 Grid Search sozlash
grid_search = GridSearchCV(
    estimator=rf,
    param_grid=param_grid,
    cv=5,                   # 5-fold cross validation
    scoring='accuracy',     # aniqlikni baholash mezoni
    n_jobs=-1,              # barcha yadroda bajarish
    verbose=2               # jarayonni ko‘rsatish
)

# 🔧 Modelni o‘qitish (eng yaxshi parametrlarni topish)
grid_search.fit(X_train, y_train)

# 🏆 Eng yaxshi model va parametrlar
print("✅ Eng yaxshi parametrlar:")
print(grid_search.best_params_)

# Eng yaxshi modelni olish
best_rf = grid_search.best_estimator_

# Bashorat
y_pred = best_rf.predict(X_test)

# 📊 Natijalar
print("\n📈 Random Forest (Grid Search) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 216 candidates, totalling 1080 fits
✅ Eng yaxshi parametrlar:
{'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 150}

📈 Random Forest (Grid Search) Accuracy: 0.5755755755755756

🔍 To‘liq tahlil:
               precision    recall  f1-score   support

         1.0       0.42      0.16      0.23       114
         2.0       0.75      0.07      0.13        42
         3.0       0.53      0.14      0.22       220
         4.0       0.58      0.93      0.71       525
         5.0       0.68      0.37      0.48        98

    accuracy                           0.58       999
   macro avg       0.59      0.33      0.35       999
weighted avg       0.57      0.58      0.50       999



#  DecisionTreeClassifier + Grid Search

In [17]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report


dt_model = DecisionTreeClassifier(random_state=42)

# 🔍 Grid Search uchun parametrlar to‘plami
param_grid = {
    'criterion': ['gini', 'entropy', 'log_loss'],  # bo‘linish mezoni
    'max_depth': [None, 5, 10, 20, 30],             # daraxt chuqurligi
    'min_samples_split': [2, 5, 10],                # tugunni bo‘lish uchun minimal namunalar
    'min_samples_leaf': [1, 2, 4],                  # barg tugun uchun eng kam namunalar
    'splitter': ['best', 'random']                  # tugun ajratish usuli
}

# 🧠 Grid Search sozlash
grid_search = GridSearchCV(
    estimator=dt_model,
    param_grid=param_grid,
    cv=5,                   # 5-fold cross-validation
    scoring='accuracy',
    n_jobs=-1,
    verbose=2
)

# 🔧 Modelni o‘qitish
grid_search.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(grid_search.best_params_)

# Eng yaxshi modelni olish
best_dt_model = grid_search.best_estimator_

# Bashorat
y_pred = best_dt_model.predict(X_test)

# 📊 Natijalar
print("\n🌳 Decision Tree (Grid Search) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 270 candidates, totalling 1350 fits
✅ Eng yaxshi parametrlar:
{'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2, 'splitter': 'best'}

🌳 Decision Tree (Grid Search) Accuracy: 0.5455455455455456

🔍 To‘liq tahlil:
               precision    recall  f1-score   support

         1.0       0.32      0.13      0.19       114
         2.0       1.00      0.05      0.09        42
         3.0       0.44      0.10      0.17       220
         4.0       0.56      0.91      0.69       525
         5.0       0.62      0.29      0.39        98

    accuracy                           0.55       999
   macro avg       0.59      0.30      0.31       999
weighted avg       0.53      0.55      0.46       999



#  Logistic Regression + Grid Search

In [18]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, classification_report


log_reg = LogisticRegression(max_iter=1000, random_state=42)

# 🔍 Grid Search uchun parametrlar to‘plami
param_grid = {
    'penalty': ['l1', 'l2', 'elasticnet', None],  # jarima turi
    'C': [0.01, 0.1, 1, 10, 100],                 # regularizatsiya kuchi
    'solver': ['lbfgs', 'liblinear', 'saga'],     # optimallashtirish algoritmi
}

# 🧠 Grid Search sozlash
grid_search = GridSearchCV(
    estimator=log_reg,
    param_grid=param_grid,
    cv=5,                   # 5-fold cross-validation
    scoring='accuracy',     # aniqlik mezoni
    n_jobs=-1,              # barcha CPU yadrolarida bajarish
    verbose=2
)

# 🔧 O‘qitish
grid_search.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(grid_search.best_params_)

# Eng yaxshi modelni olish
best_log_model = grid_search.best_estimator_

# Bashorat
y_pred = best_log_model.predict(X_test)

# 📊 Natijalar
print("\n📈 Logistic Regression (Grid Search) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))

Fitting 5 folds for each of 60 candidates, totalling 300 fits


c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\model_selection\_validation.py:516: FitFailedWarning: 
125 fits failed out of a total of 300.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
25 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\base.py", line 1365, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\

✅ Eng yaxshi parametrlar:
{'C': 0.1, 'penalty': 'l2', 'solver': 'lbfgs'}

📈 Logistic Regression (Grid Search) Accuracy: 0.5455455455455456

🔍 To‘liq tahlil:
               precision    recall  f1-score   support

         1.0       1.00      0.02      0.03       114
         2.0       0.00      0.00      0.00        42
         3.0       0.33      0.01      0.03       220
         4.0       0.54      0.98      0.70       525
         5.0       0.69      0.26      0.37        98

    accuracy                           0.55       999
   macro avg       0.51      0.25      0.23       999
weighted avg       0.54      0.55      0.41       999



c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\Rasulbek907\Desktop\Project_MP\venv\L

| № | Model nomi                           | Eng yaxshi parametrlar                                                                                          | Aniqlik (Accuracy) |
| - | ------------------------------------ | --------------------------------------------------------------------------------------------------------------- | ------------------ |
| 1 | 🌲 Random Forest (Grid Search)       | {'criterion': 'entropy', 'max_depth': None, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 150} | 0.576              |
| 2 | 🌳 Decision Tree (Grid Search)       | {'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 1, 'min_samples_split': 2, 'splitter': 'best'}     | 0.546              |
| 3 | 📊 Logistic Regression (Grid Search) | {'C': 0.1, 'penalty': 'l2', 'solver': 'lbfgs'}                                                                  | 0.546              |


# RANDOM SEARCH 

# RandomForestClasssifier + Random Search

In [22]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report
from scipy.stats import randint

# 🔹 Ma’lumotni ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train/Test bo‘lish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# 🌲 Model
rf_model = RandomForestClassifier(random_state=42)

# 🎯 Random Search uchun parametrlar diapazoni
param_dist = {
    'n_estimators': randint(50, 300),         # daraxtlar soni
    'max_depth': [None, 5, 10, 20, 30],       # chuqurlik
    'min_samples_split': randint(2, 10),      # bo‘linish uchun minimal namunalar
    'min_samples_leaf': randint(1, 5),        # barg uchun minimal namunalar
    'criterion': ['gini', 'entropy', 'log_loss'],  # ajratish mezoni
    'bootstrap': [True, False]                # bootstrap namunalar ishlatilsinmi
}

# ⚙️ Random Search sozlash
random_search = RandomizedSearchCV(
    estimator=rf_model,
    param_distributions=param_dist,
    n_iter=20,               # 20 ta tasodifiy kombinatsiyani sinaydi
    cv=5,                    # 5-fold cross-validation
    scoring='accuracy',
    random_state=42,
    n_jobs=-1,
    verbose=2
)

# 🔧 Modelni o‘qitish
random_search.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(random_search.best_params_)

# Eng yaxshi modelni olish
best_rf_model = random_search.best_estimator_

# Bashorat
y_pred = best_rf_model.predict(X_test)

# 📊 Natijalar
print("\n🌲 Random Forest (Random Search) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 20 candidates, totalling 100 fits
✅ Eng yaxshi parametrlar:
{'bootstrap': False, 'criterion': 'gini', 'max_depth': 20, 'min_samples_leaf': 4, 'min_samples_split': 9, 'n_estimators': 180}

🌲 Random Forest (Random Search) Accuracy: 0.5687583444592791

🔍 To‘liq tahlil:
               precision    recall  f1-score   support

         1.0       0.46      0.18      0.26       170
         2.0       0.80      0.06      0.12        64
         3.0       0.47      0.16      0.24       330
         4.0       0.58      0.91      0.71       787
         5.0       0.58      0.31      0.40       147

    accuracy                           0.57      1498
   macro avg       0.58      0.32      0.35      1498
weighted avg       0.55      0.57      0.50      1498



# Decision Tree Classifier + Random Search

In [23]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report
from scipy.stats import randint, uniform

# 🔹 X va y ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train/Test bo‘lish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# 🌳 Model
dt_model = DecisionTreeClassifier(random_state=42)

# 🎯 Random Search uchun parametrlar diapazoni
param_dist = {
    'criterion': ['gini', 'entropy', 'log_loss'],   # ajratish mezoni
    'splitter': ['best', 'random'],                 # bo‘linish strategiyasi
    'max_depth': [None, 5, 10, 20, 30, 50],         # daraxt chuqurligi
    'min_samples_split': randint(2, 20),            # tugunni bo‘lish uchun minimal namunalar
    'min_samples_leaf': randint(1, 10),             # barg uchun minimal namunalar
    'max_features': [None, 'sqrt', 'log2']          # xususiyat tanlash strategiyasi
}

# ⚙️ Random Search sozlash
random_search = RandomizedSearchCV(
    estimator=dt_model,
    param_distributions=param_dist,
    n_iter=20,               # 20 ta kombinatsiya sinovdan o‘tadi
    cv=5,                    # 5-fold cross-validation
    scoring='accuracy',
    random_state=42,
    n_jobs=-1,
    verbose=2
)

# 🔧 Modelni o‘qitish
random_search.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(random_search.best_params_)

# Eng yaxshi modelni olish
best_dt_model = random_search.best_estimator_

# Bashorat
y_pred = best_dt_model.predict(X_test)

# 📊 Natijalar
print("\n🌳 Decision Tree (Random Search) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 20 candidates, totalling 100 fits
✅ Eng yaxshi parametrlar:
{'criterion': 'entropy', 'max_depth': 5, 'max_features': None, 'min_samples_leaf': 1, 'min_samples_split': 13, 'splitter': 'random'}

🌳 Decision Tree (Random Search) Accuracy: 0.5393858477970628

🔍 To‘liq tahlil:
               precision    recall  f1-score   support

         1.0       0.00      0.00      0.00       170
         2.0       0.45      0.08      0.13        64
         3.0       0.08      0.00      0.01       330
         4.0       0.54      0.98      0.70       787
         5.0       0.64      0.23      0.34       147

    accuracy                           0.54      1498
   macro avg       0.34      0.26      0.24      1498
weighted avg       0.38      0.54      0.41      1498



# Logistic regression + Random Search 

In [24]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.metrics import accuracy_score, classification_report
from scipy.stats import uniform

# 🔹 X va y ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train/Test ajratish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# ⚙️ Model
log_model = LogisticRegression(max_iter=1000, random_state=42)

# 🎯 Random Search uchun parametrlar diapazoni
param_dist = {
    'solver': ['liblinear', 'lbfgs', 'saga'],  # optimallashtirish algoritmlari
    'penalty': ['l1', 'l2', 'elasticnet', None],  # regularizatsiya turi
    'C': uniform(0.01, 10),                     # regularizatsiya kuchi
    'fit_intercept': [True, False],             # interceptni o‘rganish yoki yo‘q
    'class_weight': [None, 'balanced']          # sinflarni balanslash
}

# 🧠 Random Search sozlash
random_search = RandomizedSearchCV(
    estimator=log_model,
    param_distributions=param_dist,
    n_iter=20,               # 20 ta kombinatsiya sinovdan o‘tadi
    cv=5,                    # 5-fold cross-validation
    scoring='accuracy',
    random_state=42,
    n_jobs=-1,
    verbose=2
)

# 🔧 Modelni o‘qitish
random_search.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(random_search.best_params_)

# Eng yaxshi modelni olish
best_log_model = random_search.best_estimator_

# Bashorat
y_pred = best_log_model.predict(X_test)

# 📊 Natijalar
print("\n⚙️ Logistic Regression (Random Search) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 20 candidates, totalling 100 fits
✅ Eng yaxshi parametrlar:
{'C': np.float64(6.813075385877797), 'class_weight': None, 'fit_intercept': True, 'penalty': 'l2', 'solver': 'liblinear'}

⚙️ Logistic Regression (Random Search) Accuracy: 0.5353805073431241

🔍 To‘liq tahlil:
               precision    recall  f1-score   support

         1.0       0.33      0.01      0.01       170
         2.0       0.00      0.00      0.00        64
         3.0       0.33      0.02      0.03       330
         4.0       0.54      0.99      0.70       787
         5.0       0.67      0.10      0.17       147

    accuracy                           0.54      1498
   macro avg       0.37      0.22      0.18      1498
weighted avg       0.46      0.54      0.39      1498



c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\model_selection\_validation.py:516: FitFailedWarning: 
50 fits failed out of a total of 100.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
5 fits failed with the following error:
Traceback (most recent call last):
  File "c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\model_selection\_validation.py", line 859, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
    ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\base.py", line 1365, in wrapper
    return fit_method(estimator, *args, **kwargs)
  File "c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\si

| № | Model nomi                             | Eng yaxshi parametrlar                                                                                                               | Accuracy |
| - | -------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ | -------- |
| 1 | 🌲 Random Forest (Random Search)       | {'bootstrap': False, 'criterion': 'gini', 'max_depth': 20, 'min_samples_leaf': 4, 'min_samples_split': 9, 'n_estimators': 180}       | 0.5688   |
| 2 | 🌳 Decision Tree (Random Search)       | {'criterion': 'entropy', 'max_depth': 5, 'max_features': None, 'min_samples_leaf': 1, 'min_samples_split': 13, 'splitter': 'random'} | 0.5394   |
| 3 | 📊 Logistic Regression (Random Search) | {'C': 6.813, 'class_weight': None, 'fit_intercept': True, 'penalty': 'l2', 'solver': 'liblinear'}                                    | 0.5354   |


# Bayesian Optimization

# 

# Bayesian Optimization +  RandomForestClassifier

In [26]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score, train_test_split
from skopt import BayesSearchCV
from sklearn.metrics import accuracy_score, classification_report


# 🌲 Model
rf = RandomForestClassifier(random_state=42)

# 🎯 Bayesian Optimization uchun parametrlar oralig‘i
param_space = {
    'n_estimators': (50, 300),           # daraxtlar soni
    'max_depth': (3, 30),                # chuqurlik
    'min_samples_split': (2, 10),        # tugunni bo‘lish uchun minimal namunalar
    'min_samples_leaf': (1, 5),          # barg uchun minimal namunalar
    'max_features': ['sqrt', 'log2', None], # xususiyat tanlash usuli
    'bootstrap': [True, False]           # bootstrap namunalar ishlatilsinmi
}

# ⚙️ BayesSearchCV sozlash
opt = BayesSearchCV(
    estimator=rf,
    search_spaces=param_space,
    n_iter=25,              # necha kombinatsiya sinovdan o‘tadi
    cv=5,                   # 5-fold cross-validation
    scoring='accuracy',
    n_jobs=-1,
    random_state=42,
    verbose=2
)

# 🔧 Modelni o‘qitish
opt.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(opt.best_params_)

# Eng yaxshi model
best_rf = opt.best_estimator_

# Bashorat
y_pred = best_rf.predict(X_test)

# 📊 Natijalar
print("\n🌲 Random Forest (Bayesian Optimization) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fi

# Bayesian Optimization +  DecisionTreeClassifier

In [27]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from skopt import BayesSearchCV

# 🔹 X va y ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train/Test ajratish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# 🌳 Model yaratish
dt = DecisionTreeClassifier(random_state=42)

# 🎯 Bayesian Optimization uchun parametrlar oralig‘i
param_space = {
    'criterion': ['gini', 'entropy', 'log_loss'],  # qaror mezoni
    'max_depth': (2, 30),                          # maksimal chuqurlik
    'min_samples_split': (2, 20),                  # bo‘linish uchun minimal namunalar
    'min_samples_leaf': (1, 10),                   # bargdagi minimal namunalar
    'max_features': ['sqrt', 'log2', None]         # xususiyat tanlash strategiyasi
}

# ⚙️ BayesSearchCV sozlash
opt = BayesSearchCV(
    estimator=dt,
    search_spaces=param_space,
    n_iter=25,              # sinovlar soni
    cv=5,                   # 5-fold cross-validation
    scoring='accuracy',
    n_jobs=-1,
    random_state=42,
    verbose=2
)

# 🔧 Modelni o‘qitish
opt.fit(X_train, y_train)

# 🏆 Eng yaxshi parametrlar
print("✅ Eng yaxshi parametrlar:")
print(opt.best_params_)

# Eng yaxshi modelni olish
best_dt = opt.best_estimator_

# Bashorat
y_pred = best_dt.predict(X_test)

# 📊 Natijalar
print("\n🌿 Decision Tree (Bayesian Optimization) Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fi

# Bayesian Optimization +  LogisticRegression

In [28]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from skopt import BayesSearchCV

# 🔹 X va y
X = df.drop(columns="nova_group")
y = df["nova_group"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# ⚙️ Model
log_reg = LogisticRegression(max_iter=1000, random_state=42)

# 🔧 Parametr oralig‘i (mos juftliklar bilan)
param_space = [
    ({'solver': ['liblinear'], 'penalty': ['l1', 'l2'], 'C': (1e-4, 10.0, 'log-uniform')}),
    ({'solver': ['lbfgs'], 'penalty': ['l2', None], 'C': (1e-4, 10.0, 'log-uniform')}),
    ({'solver': ['saga'], 'penalty': ['l1', 'l2', 'elasticnet', None],
      'C': (1e-4, 10.0, 'log-uniform'), 'l1_ratio': (0, 1.0)})
]

# 🚀 Bayesian Optimization
opt = BayesSearchCV(
    estimator=log_reg,
    search_spaces=param_space,
    n_iter=25,
    cv=5,
    scoring='accuracy',
    n_jobs=-1,
    random_state=42,
    verbose=2
)

opt.fit(X_train, y_train)

print("✅ Eng yaxshi parametrlar:", opt.best_params_)
best_model = opt.best_estimator_

y_pred = best_model.predict(X_test)
print("\n📈 Accuracy:", accuracy_score(y_test, y_pred))
print("\n🔍 To‘liq tahlil:\n", classification_report(y_test, y_pred))


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fits
Fitting 5 folds for each of 1 candidates, totalling 5 fi

c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 1000 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=1000).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\metrics\_classification.py:1731: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
c:\Users\Rasulbek907\Desktop\Project_MP\venv\L

| № | Model nomi                                     | Accuracy | Macro Avg (Precision) | Macro Avg (Recall) | Macro Avg (F1-score) | Weighted Avg (Precision) | Weighted Avg (Recall) | Weighted Avg (F1-score) | Support |
| - | ---------------------------------------------- | -------- | --------------------- | ------------------ | -------------------- | ------------------------ | --------------------- | ----------------------- | ------- |
| 1 | RandomForestClassifier (Bayesian Optimization) | 0.58     | 0.57                  | 0.35               | 0.39                 | 0.56                     | 0.58                  | 0.52                    | 1498    |
| 2 | DecisionTreeClassifier (Bayesian Optimization) | 0.54     | 0.49                  | 0.27               | 0.26                 | 0.52                     | 0.54                  | 0.44                    | 1498    |
| 3 | LogisticRegression (Bayesian Optimization)     | 0.54     | 0.38                  | 0.24               | 0.21                 | 0.46                     | 0.54                  | 0.40                    | 1498    |


# Optuna (Automated / Advanced Method) 

# Optuna + Random Forest Classifier

In [29]:
import optuna
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# 🔹 X va y ajratish
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train/test bo‘lish (stratify bilan)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# 🔹 Optuna maqsad funksiyasi
def objective(trial):
    n_estimators = trial.suggest_int("n_estimators", 50, 300)
    max_depth = trial.suggest_int("max_depth", 2, 30)
    min_samples_split = trial.suggest_int("min_samples_split", 2, 10)
    min_samples_leaf = trial.suggest_int("min_samples_leaf", 1, 10)
    max_features = trial.suggest_categorical("max_features", ["sqrt", "log2", None])

    model = RandomForestClassifier(
        n_estimators=n_estimators,
        max_depth=max_depth,
        min_samples_split=min_samples_split,
        min_samples_leaf=min_samples_leaf,
        max_features=max_features,
        random_state=42,
        n_jobs=-1
    )

    # 3-fold cross-validation orqali aniqlikni o‘lchash
    score = cross_val_score(model, X_train, y_train, cv=3, scoring="accuracy").mean()
    return score

# 🔹 Study yaratish
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=30)

# 🔹 Natijalar
print("🥇 Eng yaxshi parametrlar:")
print(study.best_params)
print(f"✅ Eng yaxshi aniqlik (CV): {study.best_value:.4f}")

# 🔹 Eng yaxshi modelni yaratish va testda sinovdan o‘tkazish
best_model = RandomForestClassifier(**study.best_params, random_state=42)
best_model.fit(X_train, y_train)
y_pred = best_model.predict(X_test)

print(f"🎯 Test aniqligi: {accuracy_score(y_test, y_pred):.4f}")


[I 2025-11-04 10:14:58,261] A new study created in memory with name: no-name-78b6b4cb-601f-4037-a2af-9e8ff87de8f4
[I 2025-11-04 10:14:59,264] Trial 0 finished with value: 0.5442060085836911 and parameters: {'n_estimators': 179, 'max_depth': 7, 'min_samples_split': 6, 'min_samples_leaf': 9, 'max_features': 'log2'}. Best is trial 0 with value: 0.5442060085836911.
[I 2025-11-04 10:14:59,646] Trial 1 finished with value: 0.5413447782546496 and parameters: {'n_estimators': 58, 'max_depth': 6, 'min_samples_split': 3, 'min_samples_leaf': 4, 'max_features': 'log2'}. Best is trial 0 with value: 0.5442060085836911.
[I 2025-11-04 10:15:01,020] Trial 2 finished with value: 0.5444921316165952 and parameters: {'n_estimators': 214, 'max_depth': 7, 'min_samples_split': 6, 'min_samples_leaf': 4, 'max_features': 'log2'}. Best is trial 2 with value: 0.5444921316165952.
[I 2025-11-04 10:15:02,661] Trial 3 finished with value: 0.5490701001430615 and parameters: {'n_estimators': 251, 'max_depth': 25, 'min_s

🥇 Eng yaxshi parametrlar:
{'n_estimators': 234, 'max_depth': 22, 'min_samples_split': 7, 'min_samples_leaf': 3, 'max_features': None}
✅ Eng yaxshi aniqlik (CV): 0.5597
🎯 Test aniqligi: 0.5821


# Optuna + Decision Tree  Classifier

In [30]:
import optuna
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score

# 🔹 X va y
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train / Test bo‘lish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# 🔹 Optuna maqsad funksiyasi
def objective(trial):
    # Parametrlarni tanlash
    criterion = trial.suggest_categorical("criterion", ["gini", "entropy", "log_loss"])
    max_depth = trial.suggest_int("max_depth", 2, 30)
    min_samples_split = trial.suggest_int("min_samples_split", 2, 20)
    min_samples_leaf = trial.suggest_int("min_samples_leaf", 1, 10)
    splitter = trial.suggest_categorical("splitter", ["best", "random"])

    # Model yaratish
    model = DecisionTreeClassifier(
        criterion=criterion,
        max_depth=max_depth,
        min_samples_split=min_samples_split,
        min_samples_leaf=min_samples_leaf,
        splitter=splitter,
        random_state=42
    )

    # Cross-validation orqali baholash
    score = cross_val_score(model, X_train, y_train, cv=3, scoring="accuracy").mean()
    return score

# 🔹 Optuna study yaratish
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=30)

# 🔹 Natijalarni chiqarish
print("🌲 Eng yaxshi parametrlar:")
print(study.best_params)
print(f"✅ Eng yaxshi aniqlik (CV): {study.best_value:.4f}")

# 🔹 Eng yaxshi modelni testda sinash
best_model = DecisionTreeClassifier(**study.best_params, random_state=42)
best_model.fit(X_train, y_train)
y_pred = best_model.predict(X_test)

print(f"🎯 Test aniqligi: {accuracy_score(y_test, y_pred):.4f}")


[I 2025-11-04 10:16:07,370] A new study created in memory with name: no-name-9811e2b0-b359-453b-aa2f-c2dc3532dc8d
[I 2025-11-04 10:16:07,463] Trial 0 finished with value: 0.48440629470672386 and parameters: {'criterion': 'gini', 'max_depth': 11, 'min_samples_split': 11, 'min_samples_leaf': 8, 'splitter': 'best'}. Best is trial 0 with value: 0.48440629470672386.
[I 2025-11-04 10:16:07,555] Trial 1 finished with value: 0.4663805436337625 and parameters: {'criterion': 'entropy', 'max_depth': 17, 'min_samples_split': 5, 'min_samples_leaf': 9, 'splitter': 'best'}. Best is trial 0 with value: 0.48440629470672386.
[I 2025-11-04 10:16:07,643] Trial 2 finished with value: 0.4666666666666666 and parameters: {'criterion': 'entropy', 'max_depth': 11, 'min_samples_split': 3, 'min_samples_leaf': 2, 'splitter': 'best'}. Best is trial 0 with value: 0.48440629470672386.
[I 2025-11-04 10:16:07,704] Trial 3 finished with value: 0.46151645207439196 and parameters: {'criterion': 'gini', 'max_depth': 29, 'm

🌲 Eng yaxshi parametrlar:
{'criterion': 'log_loss', 'max_depth': 3, 'min_samples_split': 16, 'min_samples_leaf': 1, 'splitter': 'random'}
✅ Eng yaxshi aniqlik (CV): 0.5416
🎯 Test aniqligi: 0.5407


# Optuna + Logistic Regression Classifier

In [33]:
import optuna
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score

# 🔹 X va y
X = df.drop(columns="nova_group")
y = df["nova_group"]

# 🔹 Train / Test bo‘lish
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42, stratify=y
)

# 🔹 Optuna maqsad funksiyasi
def objective(trial):
    # Faqat C va solver parametrlarini optimallashtiramiz
    C = trial.suggest_float("C", 0.01, 10.0, log=True)
    solver = trial.suggest_categorical("solver", ["lbfgs", "liblinear", "saga", "newton-cg"])

    # Model yaratish
    model = LogisticRegression(
        C=C,
        solver=solver,
        max_iter=1000,
        random_state=42
    )

    # Cross-validation orqali baholash
    score = cross_val_score(model, X_train, y_train, cv=3, scoring="accuracy").mean()
    return score

# 🔹 Optuna study yaratish
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=30)

# 🔹 Natijalarni chiqarish
print("🌟 Eng yaxshi parametrlar:")
print(study.best_params)
print(f"✅ Eng yaxshi aniqlik (CV): {study.best_value:.4f}")

# 🔹 Eng yaxshi modelni testda sinash
best_model = LogisticRegression(**study.best_params, max_iter=1000, random_state=42)
best_model.fit(X_train, y_train)
y_pred = best_model.predict(X_test)

print(f"🎯 Test aniqligi: {accuracy_score(y_test, y_pred):.4f}")

[I 2025-11-04 10:19:07,329] A new study created in memory with name: no-name-88bf179e-9d0a-4c9d-893b-93a2d2ca5a45


[I 2025-11-04 10:19:08,208] Trial 0 finished with value: 0.5402002861230328 and parameters: {'C': 4.28949261590443, 'solver': 'newton-cg'}. Best is trial 0 with value: 0.5402002861230328.
[I 2025-11-04 10:19:09,011] Trial 1 finished with value: 0.5402002861230328 and parameters: {'C': 3.945542549111091, 'solver': 'newton-cg'}. Best is trial 0 with value: 0.5402002861230328.
c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
c:\Users\Rasulbek907\Desktop\Project_MP\venv\Lib\site-packages\sklearn\linear_model\_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
[I 2025-11-04 10:19:11,672] T

🌟 Eng yaxshi parametrlar:
{'C': 8.81401263866425, 'solver': 'newton-cg'}
✅ Eng yaxshi aniqlik (CV): 0.5402
🎯 Test aniqligi: 0.5401


| № | Model nomi                       | Eng yaxshi parametrlar                                                                                  | Eng yaxshi aniqlik (CV) | Test aniqligi |
|---|---------------------------------|--------------------------------------------------------------------------------------------------------|-------------------------|---------------|
| 1 | Random Forest Classifier (Optuna) | {'n_estimators': 234, 'max_depth': 22, 'min_samples_split': 7, 'min_samples_leaf': 3, 'max_features': None} | 0.5597                  | 0.5821        |
| 2 | Decision Tree Classifier (Optuna) | {'criterion': 'log_loss', 'max_depth': 3, 'min_samples_split': 16, 'min_samples_leaf': 1, 'splitter': 'random'} | 0.5416                  | 0.5407        |
| 3 | Logistic Regression (Optuna)      | {'C': 8.81401263866425, 'solver': 'newton-cg'}                                                       | 0.5402                  | 0.5401        |
